### PR TITLE
gh-94215: Fix reference count issue in exception_unwind

### PR DESCRIPTION
**Pull request by tiran**
Zero-cost exception handling did not take ``frame_setlineno()`` into
account. It can pop off stacks. This can lead to a crash.

Exception unwinding now re-calculates the stack pointer.

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

* Issue: gh-94215

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2022-07-07-17-07-00.gh-issue-94215.cXltGH.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Fix a reference counting bug in exception unwinding code for zero-cost
+exception handling. The code did not take into account that
+``frame_setlineno()`` can pop off stacks, too.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2022-07-07-17-07-00.gh-issue-94215.cXltGH.rst:3]:
```suggestion
``frame_setlineno()`` can pop items off the stack, too.
```

Comment (tiran) [reply Misc/NEWS.d/next/Core and Builtins/2022-07-07-17-07-00.gh-issue-94215.cXltGH.rst:3]: Thanks! I'll apply the grammar fix after buildbots are done with testing.

**Comment by tiran** [comment]
The fix seems to introduce a reference leak:

```
test_sys_settrace leaked [17, 17, 17] references, sum=51
test_sys_setprofile leaked [4, 4, 4] references, sum=12
```

**Comment by encukou** [comment]
The leaking tests in `test_sys_settrace` are:
```
test_no_jump_between_except_blocks
test_no_jump_within_except_block
test_no_jump_to_except_1
test_no_jump_to_except_2
test_no_jump_to_except_3
test_no_jump_to_except_4
test_no_jump_between_except_blocks_2
test_no_jump_into_bare_except_block_from_try_block
test_no_jump_into_qualified_except_block_from_try_block
test_no_jump_out_of_bare_except_block
test_no_jump_out_of_qualified_except_block
```

File: Python/ceval.c [hunk 2]
```diff
@@ -5795,6 +5795,11 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 
                 /* Pop remaining stack entries. */
                 PyObject **stackbase = _PyFrame_Stackbase(frame);
+                if (frame->stacktop != -1) {
```
Comment (ericsnowcurrently) [inline Python/ceval.c]: Hmm, this has a smell to it.  Why is this necessary when it isn't necessary for any other use of `_PyFrame_GetStackPointer()`?  This makes me think we're missing something.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

At the least this seems to be an effective short-term solution.  It may be valid for the long-term too.  I've left some thoughts on [the issue](https://github.com/python/cpython/issues/94215#issuecomment-1178073348).

File: Python/ceval.c (hunk 2 above)
Comment (tiran) [reply Python/ceval.c]: I agree. I don't understand the code well enough. All I can say is that the code was crashing without the check.

Comment (ericsnowcurrently) [reply Python/ceval.c]: We can follow up with @markshannon when he's available again.

**Comment by brandtbucher** [comment]
@tiran, applying this diff to your branch fixes the leaks for me:

```diff
diff --git a/Python/ceval.c b/Python/ceval.c
index 083f881807..e23bfda725 100644
--- a/Python/ceval.c
+++ b/Python/ceval.c
@@ -5683,6 +5683,8 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
                     err = maybe_call_line_trace(tstate->c_tracefunc,
                                                 tstate->c_traceobj,
                                                 tstate, frame, instr_prev);
+                    stack_pointer = _PyFrame_GetStackPointer(frame);
+                    frame->stacktop = -1;
                     if (err) {
                         /* trace function raised an exception */
                         next_instr++;
@@ -5690,9 +5692,6 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
                     }
                     /* Reload possibly changed frame fields */
                     next_instr = frame->prev_instr;
-
-                    stack_pointer = _PyFrame_GetStackPointer(frame);
-                    frame->stacktop = -1;
                 }
             }
         }
@@ -5795,11 +5794,6 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 
                 /* Pop remaining stack entries. */
                 PyObject **stackbase = _PyFrame_Stackbase(frame);
-                if (frame->stacktop != -1) {
-                    // frame_setlineno() may have popped off additional stacks in
-                    // frame_stack_pop(). Re-calculate stack pointer.
-                    stack_pointer = _PyFrame_GetStackPointer(frame);
-                }
                 while (stack_pointer > stackbase) {
                     PyObject *o = POP();
                     Py_XDECREF(o);
```

I'm not entirely sure *why* yet, but my understanding (gained over the last couple of hours) is that `frame->stacktop` is expected to always be `-1` in this function, *except* when we're calling a trace function (since GC of frame objects needs to work correctly there). This looks like one place where we forget to set it back to `-1` and reload the (possibly changed) stack pointer if an error is raised in while tracing...

...maybe?

File: Python/ceval.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Python/ceval.c]: Looks like @brandtbucher figured out the smell. 😄
